Woodoku/tests/entity/test_woodoku_board.py

```diff
@@ -0,0 +1,163 @@
+        (gun_shape, (6, 0)),
+    ])
+    def test_can_add_shape_at_location_on_empty_board_within_board(self, shape: WoodokuShape,
+                                                                   location: Tuple[int, int]) -> None:
+        """
+        Add shape to an empty board where each location to add is at the edge of the game board
+        """
+        board = WoodokuBoard()
+        assert board.can_add_shape_at_location(shape, *location)
+
+    @pytest.mark.parametrize('shape, location', [
+        (l_shape, (6, 8)),
+        (horizontal_bar_shape, (4, 3)),
+        (gun_shape, (4, 0)),
+    ])
+    def test_can_add_shape_at_location_on_empty_board_go_beyond_board(self, shape: WoodokuShape,
+                                                                      location: Tuple[int, int]) -> None:
+        """
+        Add shape to an empty board where location goes beyond the game board
+        """
+        board = WoodokuBoard()
+        with pytest.raises(ShapeOutOfBoardError):
+            board.can_add_shape_at_location(shape, *location)
+
+    @pytest.mark.parametrize('first_shape, first_position, sec_shape, sec_position', [
+        (l_shape, (0, 0), l_shape, (1, 0)),
+        (horizontal_bar_shape, (4, 3), gun_shape, (4, 4)),
+        (gun_shape, (4, 0), horizontal_bar_shape, (5, 0)),
+    ])
+    def test_can_add_shape_at_occupied_location(self, first_shape: WoodokuShape, first_position: Tuple[int, int],
+                                                sec_shape: WoodokuShape, sec_position: Tuple[int, int]) -> None:
+        """
+        first_shape and first_location are chosen so that sec_shape will be overlapped with first_shape when
+        adding it at sec_location
+        """
+        board = WoodokuBoard()
+        board.add_shape(first_shape, *first_position)
+        assert not board.can_add_shape_at_location(sec_shape, *sec_position)
+
+    @pytest.mark.parametrize('shape', [
+        l_shape,
+        horizontal_bar_shape,
+        gun_shape
+    ])
+    def test_can_add_shape_to_board_on_empty_board(self, shape: WoodokuShape) -> None:
+        board = WoodokuBoard()
+        assert board.can_add_shape_to_board(shape)
+
+    def test_can_add_shape_to_board_on_conflicted_board(self) -> None:
+        """
+        first_shape and first_location are chosen so that sec_shape will be overlapped with first_shape when
+        adding it at sec_location
+        """
+        board = WoodokuBoard()
+        for i in range(9):
+            assert board.can_add_shape_to_board(self.horizontal_bar_shape)
+        assert not board.can_add_shape_to_board(self.horizontal_bar_shape)
+
+    def test_find_groups_a_row(self) -> None:
+        """
+        A first row should be the only group on the board
+        """
+        board = WoodokuBoard()
+        board.add_shape(self.horizontal_bar_shape, 0, 0)
+        board.add_shape(self.gun_shape, 0, 5)
+        board.add_shape(self.vertical_two_block, 0, 8)
```

Comment:
Theoretically, after this last `add_shape`, this row is already removed. Thus the `_WoodokuBoard__find_groups` will not show this row. 

I would suggest to manipulate `_WoodokuBoardRepresentation` within `WoodokuBoard` directly to test `_WoodokuBoard__find_groups`.